tests/test_isPrimaryFlag.py

```diff
@@ -36,31 +176,120 @@ def setUp(self):
         # Load sample input from disk
         expPath = os.path.join(getPackageDir("pipe_tasks"), "tests", "data", "v695833-e0-c000-a00.sci.fits")
         self.exposure = afwImage.ExposureF(expPath)
+
+        # Characterize the image (create PSF, etc.)
+        charImConfig = CharacterizeImageConfig()
+        charImTask = CharacterizeImageTask(config=charImConfig)
+        self.charImResults = charImTask.run(self.exposure)
+
         # set log level so that warnings do not display
         Log.getLogger("calibrate").setLevel(Log.ERROR)
 
     def tearDown(self):
         del self.exposure
 
-    def testIsPrimaryFlag(self):
+    def testIsSinglePrimaryFlag(self):
         """Tests detect_isPrimary column gets added when run, and that sources
         labelled as detect_isPrimary are not sky sources and have no children.
         """
-        charImConfig = CharacterizeImageConfig()
-        charImTask = CharacterizeImageTask(config=charImConfig)
-        charImResults = charImTask.run(self.exposure)
         calibConfig = CalibrateConfig()
         calibConfig.doAstrometry = False
         calibConfig.doPhotoCal = False
         calibTask = CalibrateTask(config=calibConfig)
-        calibResults = calibTask.run(charImResults.exposure)
+        calibResults = calibTask.run(self.charImResults.exposure)
         outputCat = calibResults.outputCat
         self.assertTrue("detect_isPrimary" in outputCat.schema.getNames())
         # make sure all sky sources are flagged as not primary
         self.assertEqual(sum((outputCat["detect_isPrimary"]) & (outputCat["sky_source"])), 0)
         # make sure all parent sources are flagged as not primary
         self.assertEqual(sum((outputCat["detect_isPrimary"]) & (outputCat["deblend_nChild"] > 0)), 0)
 
+    def testIsScarletPrimaryFlag(self):
```

Comment:
This should be cleaned up in the new commit. The previous version was using an older version of the mock skymap, tract, etc.

Reply:
Ok, thanks.  Now that you have the functionality, would you mind adding a similar test for `meas_deblender` runs and, in particular, making sure any `scarlet`-specific columns (whatever they be named) are NOT present?  I.e. something like (warning: not fully tested):
```
    def testIsMeasDeblenderPrimaryFlag(self):
        """Test detect_isPrimary column for meas_deblender runs.
        """
        coadd = self.exposure
        # Create a SkyMap with a tract that contains a portion of the image,
        # subdivided into 3x3 patches.
        wcs = coadd.getWcs()
        tractBBox = Box2I(Point2I(100, 100), Extent2I(900, 900))
        skyMap = MockSkyMap([tractBBox], wcs, 3)
        tractInfo = skyMap[0]
        patchInfo = tractInfo[0, 0]
        patchBBox = patchInfo.getInnerBBox()

        schema = SourceCatalog.Table.makeMinimalSchema()
        # Initialize the detection task
        detectionTask = SourceDetectionTask(schema=schema)

        # Initialize the fake source injection task
        skyConfig = SkyObjectsTask.ConfigClass()
        skySourcesTask = SkyObjectsTask(name="skySources", config=skyConfig)
        schema.addField("merge_peak_sky", type="Flag")

        # Initialize the deblender task
        deblendTask = SourceDeblendTask(schema=schema)

        # We'll customize the configuration of measurement to just run the
        # minimal number of plugins to make setPrimaryFlags work.
        measureConfig = SingleFrameMeasurementTask.ConfigClass()
        measureConfig.plugins.names = ["base_SdssCentroid", "base_SkyCoord"]
        measureConfig.slots.psfFlux = None
        measureConfig.slots.apFlux = None
        measureConfig.slots.shape = None
        measureConfig.slots.modelFlux = None
        measureConfig.slots.calibFlux = None
        measureConfig.slots.gaussianFlux = None
        measureTask = SingleFrameMeasurementTask(config=measureConfig, schema=schema)
        primaryConfig = SetPrimaryFlagsTask.ConfigClass()
        setPrimaryTask = SetPrimaryFlagsTask(config=primaryConfig, schema=schema,
                                             name="setPrimaryFlags", isSingleFrame=False)
        table = SourceCatalog.Table.make(schema)
        # detect sources
        detectionResult = detectionTask.run(table=table, exposure=coadd)
        catalog = detectionResult.sources
        # add fake sources
        skySources = skySourcesTask.run(mask=coadd.mask, seed=0)
        for foot in skySources[:5]:
            src = catalog.addNew()
            src.setFootprint(foot)
            src.set("merge_peak_sky", True)
        # deblend
        deblendTask.run(exposure=coadd, sources=catalog)
        catalog = catalog.copy(deep=True)
        measureTask.run(measCat=catalog, exposure=coadd)
        outputCat = catalog.copy(deep=True)
        # Set the primary flags
        setPrimaryTask.run(outputCat, skyMap=skyMap, tractInfo=tractInfo, patchInfo=patchInfo)

        # Check that SCARLET-only columns are NOT in the schema
        self.assertTrue("SCARLET_NAME" not in outputCat.schema.getNames())

        # Check that the sources contained in a tract are all marked
        # appropriately.
        x = outputCat["slot_Centroid_x"]
        y = outputCat["slot_Centroid_y"]
        tractInner = tractBBox.contains(x, y)
        # tractInner = bboxContainsColumn(tractBBox, x, y)
        np.testing.assert_array_equal(outputCat["detect_isTractInner"], tractInner)

        # Check that the sources contained in a patch are all marked
        # appropriately.
        patchInner = bboxContainsColumn(patchBBox, x, y)
        np.testing.assert_array_equal(outputCat["detect_isPatchInner"], patchInner)

        # make sure all sky sources are flagged as not primary
        self.assertEqual(sum((outputCat["detect_isPrimary"]) & (outputCat["merge_peak_sky"])), 0)
```

Reply:
I don't think that this is necessary, since `meas_deblender` (`SourceDeblendTask`) is run in `testIsSinglePrimaryFlag`. I can add some checks to make sure that the scarlet columns are not contained in the outputs, but I don't think that this test will touch any code that isn't covered  by the two existing tests. Please correct me if I missed something.

Reply:
It may seem like slight overkill, but `testIsSinglePrimary` above is for a single frame-like catalog (no tract or patch, different sky column name...different code path to setting the flags).  It actually wasn’t tested before because no one had gone to the trouble of mocking up a skymap...but you did, so I’m taking full advantage of your efforts (but wrote the above myself in hopes of sparing you too much extra work for your efforts!)

Reply:
I think that my latest push corrected everything but this. It isn't that it's an inconvenience to implement this test, it would be trivial to implement a deblender generic function and call it once with `ScarletDeblendTask` and once with `SourceDeblendTask`. My issue with this request is that it doesn't actually test any new code paths but will add to the overall runtime of the already long pipe_tasks tests. All of the deblend flags are logically separate from the tract/patch code blocks, and looking back through the code I am fairly confident that this module is fully covered without it.

Reply:
Yes, either I'm not understanding you or you're not understanding me, but I'll give it one last shot.

You are correct that when meas_deblender is run, the flags set by deblending are not tested with the flags from a skymap. So if the results of the deblender were dependent on those flags in any way, then not testing that would be missing a possible source of failure. Or, if the flags set by the deblender or isPrimary were in any way correlated then this would also be something that needs to be tested.

_But_... the deblender is completely isolated and independent of those flags. Similarly, the creation of the deblender flags is completely independent of the patch/tract flags. So it doesn't matter _which_ deblender the patch/tract flags are tested against, as long as they are tested against a set of deblender flags, which they are. I'm fairly certain that coverage of this module is 100%, but if you can find a code path that isn't tested then please let me know.

Reply:
Ok, I'm convinced this can be left as is!